### ISSUE TITLE
rust-mode cannot be installed due to `File is missing: treesit` error

### ISSUE THREAD

**Issue by mario-anchor** [doomemacs/doomemacs#7698]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.
- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.
- [X] The issue can be reproduced on the **latest** available commit of Doom.
- [X] The issue can be reproduced on a stable release of Emacs, such as 27, 28, or 29. *(Unstable versions end in .50, .60, or .9x)*

### Expected behavior

Running `doom sync -u` should work

### Current behavior

When I do `doom sync` or `doom sync -u` I get the following:

```
$ doom sync -u
! Failed to locate a ':tools gist' module
> Synchronizing "default" profile...
  > Regenerating envvars file
    ✓ Generated ~/.config/emacs/.local/env
  x There was an unexpected runtime error
    Message: File is missing
    Details: ("Cannot open load file" "No such file or directory" "treesit")
    Backtrace:
      (require treesit)
      (eval-buffer #<buffer  *load*-317521> nil "/home/mario_camou_anchorlabs_com/.config/emacs/.local/straight/build-27.1/rust-mode/...
      (load-with-code-conversion "/home/mario_camou_anchorlabs_com/.config/emacs/.local/straight/build-27.1/rust-mode/rust-mode-autol...
      (load "/home/mario_camou_anchorlabs_com/.config/emacs/.local/straight/build-27.1/rust-mode/rust-mode-autoloads.el" nil nomessage)
      (progn (load autoloads-file nil 'nomessage))
      (if (file-exists-p autoloads-file) (progn (load autoloads-file nil 'nomessage)))
      (let ((autoloads-file (straight--autoloads-file package))) (if (file-exists-p autoloads-file) (progn (load autoloads-file nil '...
      (straight--load-package-autoloads "rust-mode")
      (let ((name (symbol-name package))) (add-to-list 'load-path (directory-file-name (straight--build-dir name))) (straight--load-p...
      (let ((package (car --dolist-tail--))) (straight-register-package package) (let ((name (symbol-name package))) (add-to-list 'lo...
      (while --dolist-tail-- (let ((package (car --dolist-tail--))) (straight-register-package package) (let ((name (symbol-name pack...
      (let ((--dolist-tail-- (cl-delete-duplicates packages :test #'equal))) (while --dolist-tail-- (let ((package (car --dolist-tail...
    ! Wrote extended backtrace to ~/.config/emacs/.local/state/logs/cli.doom.240227104929.38480.error
  ! Script was abruptly aborted, leaving Doom in an incomplete state!
  - Run 'doom sync' to repair it.
```

### Steps to reproduce

1. Enable `:lang rust +lsp`
2. Run `doom sync -u` (it actually started after `doom upgrade`)

### System Information

https://pastebin.com/HXQ8D5Jv

**Comment by peterhoeg** [comment]
looks like it expects `treesit` which is only available in emacs 29 and above. Are you able to update emacs?

**Comment by mcamou** [comment]
No, I have to use Emacs 27.1 which comes with Debian Bullseye (company policy)

**Comment by mario-anchor** [comment]
No, I have to use Emacs 27.1 which comes with Debian 11 Bullseye (company policy). It seems that for pre-29 you have to install the tree-sitter package: https://github.com/rust-lang/rust-mode#tree-sitter. I tried adding (package! tree-sitter) in .config/doom/packages.el, but I'm still getting the error.

**Comment by peterhoeg** [comment]
This isn't a doom issue but that `rust-mode` changed how it handles tree-sitter support. Have a look at https://github.com/rust-lang/rust-mode?tab=readme-ov-file#tree-sitter.

Even if you have to use debian 12, maybe compiling a more recent emacs on that box is an option and sticking it somewhere under $HOME? There are *many* great QoL improvements since emacs 27.

**Comment by peterhoeg** [comment]
You can also pin an older `rust-mode` of course in your doom config.

**Comment by hlissner** [comment]
This is caused by https://github.com/rust-lang/rust-mode/commit/08cea61390971d26b982677bc0658a9ff79b34e0, which autoloads `treesit`, forcing it to be loaded at startup (when autoloads are initialized) unconditionally. I'll look into pinning `rust-mode` itself in the rust module.

In the meantime, see if this fixes the issue for you:

```elisp
;;; add to $DOOMDIR/packages.el
(package! rust-mode :pin "f74dd1cd87987ea7faf0cfc6240c2284ef9133cb")
```

**Comment by TheGitHub428Guy** [comment]
There's a new commit at https://github.com/rust-lang/rust-mode/commit/a8f94697acbd9d71ec94b79a30685899e9c31c6a which (I think) fixes this issue.

For anyone else trying to switch to a different commit but still getting the `treesit`-related error, here are the steps that worked for me:

1. In `$DOOMDIR/init.el`, comment out the line containing `(rust +lsp)`.
   - Make sure to also comment out other things in your config that use `rust-mode`, if any.
2. Run `doom purge -g` to get rid of the old `rust-mode`.
3. Add the line mentioned in the previous post to `$DOOMDIR/packages.el`. If you want to use the new commit, change the pin to `"a8f94697acbd9d71ec94b79a30685899e9c31c6a"`.
   - I'm not actually sure if this step is necessary (I'm not familiar with how/which commits are chosen by Doom), but I did it just to be safe
4. Uncomment the stuff you commented out in step 1 and run `doom sync -u`.

**Comment by ak-1** [comment]
> (package! rust-mode :pin "f74dd1cd87987ea7faf0cfc6240c2284ef9133cb")

This does fix the issue for me, thanks!

**Comment by mario-anchor** [comment]
> There's a new commit at [rust-lang/rust-mode@a8f9469](https://github.com/rust-lang/rust-mode/commit/a8f94697acbd9d71ec94b79a30685899e9c31c6a) which (I think) fixes this issue.
> 
> For anyone else trying to switch to a different commit but still getting the `treesit`-related error, here are the steps that worked for me:
> 
> 1. In `$DOOMDIR/init.el`, comment out the line containing `(rust +lsp)`.
>    
>    * Make sure to also comment out other things in your config that use `rust-mode`, if any.
> 2. Run `doom purge -g` to get rid of the old `rust-mode`.
> 3. Add the line mentioned in the previous post to `$DOOMDIR/packages.el`. If you want to use the new commit, change the pin to `"a8f94697acbd9d71ec94b79a30685899e9c31c6a"`.
>    
>    * I'm not actually sure if this step is necessary (I'm not familiar with how/which commits are chosen by Doom), but I did it just to be safe
> 4. Uncomment the stuff you commented out in step 1 and run `doom sync -u`.

This worked for me but I had to use commit `8b1ef9f` (currently latest  `master`) instead of `a8f9469`.

**Comment by hlissner** [comment]
rust-lang/rust-mode@08cea6139097 causes rust-mode to eagerly load those packages at startup/when the package is installed, which is a glaring oversight that needs to be dealt with upstream.

That said, I've now pinned rust-mode to a commit that at least solves the file-missing error. Thanks for bringing it to my attention!

**Comment by mcamou** [comment]
@hlissner thanks for the fix! I pinned `rust-mode` to `8b1ef9f` (which is currently latest `master`) and it's fixed, so you might want to unpin again at some point.
